Sharded DDP Docs (#4920)
PyTorchLightning/pytorch-lightning: docs/source/multi_gpu.rst

    cd /project/src
    python some_file.py --accelerator 'ddp' --gpus 8

> Sharded DDP

Reviewer: ```suggestion
Sharded DDP [BETA]
```

^^^^^^^^^^^
Lightning integration of `DeepSpeed ZeRO <https://arxiv.org/abs/1910.02054>`_ and
`ZeRO-2 <https://www.microsoft.com/en-us/research/blog/zero-2-deepspeed-shattering-barriers-of-deep-learning-speed-scale/>`_
provided by `Fairscale <https://github.com/facebookresearch/fairscale/tree/master/fairscale>`_. Sharded DDP is similar to normal DDP, except optimizer state and gradients are sharded across your GPUs.

> This means the memory overhead per GPU is less, as each GPU only has to maintain a section of your optimizer state and gradients.

Reviewer: ```suggestion
This means the memory overhead per GPU is lower, as each GPU only has to maintain a section of your optimizer state and gradients.
```

The benefits are variable by model, but we've recorded up to a 25% memory reduction per GPU. Because of extremely efficient communication,
these benefits in multi-GPU setups are almost free and throughput scales well with multi-node setups.

It is highly recommended to use Sharded DDP in multi-GPU environments where memory is limited or where training larger models are beneficial.

> Before running, install Fairscale using the command below or install all extras using ``pip install pytorch-lightning["extra"]``.

Reviewer: ```suggestion
To use shared DDP, you need to first install Fairscale using the command below or install all extras using ``pip install pytorch-lightning["extra"]``.
```

    cd /project/src
    python some_file.py --accelerator 'ddp' --gpus 8

> 

Reviewer: rmove?

----------

Model Parallelism [BETA]
------------------------

Model Parallelism tackles training large models on distributed systems, by modifying distributed communications and memory management of the model.
Unlike data parallelism, the model is partitioned in various ways across the GPUs, in most cases to reduce the memory overhead when training large models.
This is useful when dealing with large Transformer based models, or in environments where GPU memory is limited.

Lightning currently offers the following methods to leverage model parallelism:

- Optimizer Sharded Training (partitioning your gradients and optimizer state across multiple GPUs, for reduced memory overhead)

Optimizer Sharded Training
^^^^^^^^^^^^^^^^^^^^^^^^^^
Lightning integration of optimizer sharded training provided by `Fairscale <https://github.com/facebookresearch/fairscale>`_.
The technique can be found within `DeepSpeed ZeRO <https://arxiv.org/abs/1910.02054>`_ and
`ZeRO-2 <https://www.microsoft.com/en-us/research/blog/zero-2-deepspeed-shattering-barriers-of-deep-learning-speed-scale/>`_,
however the implementation is built from the ground up to be pytorch compatible and standalone.

> Optimizer Sharded Training still utilizes Data Parallel Training under the hood, except optimizer state and gradients are sharded across GPUs.

Reviewer: ```suggestion
Optimizer Sharded Training still utilizes Data Parallel Training under the hood, except the optimizer state and gradients which are sharded across GPUs.
```

> This means the memory overhead per GPU is lower, as each GPU only has to maintain a section of your optimizer state and gradients.

Reviewer: ```suggestion
This means the memory overhead per GPU is lower, as each GPU only has to maintain a partition of your optimizer state and gradients.
```

> The benefits are variable by model, but we've recorded up to a 63% memory reduction per GPU allowing us to double our model sizes. Because of extremely efficient communication,

Reviewer: maybe vary? not sure

these benefits in multi-GPU setups are almost free and throughput scales well with multi-node setups.

It is highly recommended to use Optimizer Sharded Training in multi-GPU environments where memory is limited, or where training larger models are beneficial (rough minimum of 500+ million parameter models).
Optimizer Sharded Training is typically not suited for smaller models, or where large batch sizes are important.
This is primarily because with larger batch sizes, storing activations for the backwards pass becomes the bottleneck in training. Sharding optimizer state as a result becomes less impactful.

To use Optimizer Sharded Training, you need to first install Fairscale using the command below or install all extras using ``pip install pytorch-lightning["extra"]``.

.. code-block:: bash

    pip install https://github.com/facebookresearch/fairscale/archive/master.zip

.. code-block:: python

    # train using Sharded DDP
    trainer = Trainer(accelerator='ddp', plugins='ddp_sharded')

Optimizer Sharded Training can work across all DDP variants by adding the additional ``--plugins ddp_sharded`` flag.

> Internally we re-initialize optimizers, sharding the optimizers across your machines and processes. We handle all communication using PyTorch distributed, so no code changes are required.

Reviewer 2: ```suggestion
Internally we re-initialize your optimizers and shard them across your machines and processes. We handle all communication using PyTorch distributed, so no code changes are required.
```

----------

> Model Parallelism [BETA]

Reviewer: Maybe above batch size?

------------------------

Model Parallelism tackles training large models on distributed systems, by modifying distributed communications and memory management of the model.
Unlike data parallelism, the model is partitioned in various ways across the GPUs, in most cases to reduce the memory overhead.
This is useful when dealing with large Transformer based models, or in environments where GPU memory is limited.

> Lightning currently offers these methods to leverage model parallelism.

Reviewer: ```suggestion
Lightning currently offers the following methods to leverage model parallelism:
```

- Optimizer Sharded Training (partitioning your gradients and optimizer state across multiple GPUs, for reduced memory overhead)

Optimizer Sharded Training
^^^^^^^^^^^^^^^^^^^^^^^^^^
Lightning integration of optimizer sharded training provided by `Fairscale <https://github.com/facebookresearch/fairscale>`_.
The technique can be found within `DeepSpeed ZeRO <https://arxiv.org/abs/1910.02054>`_ and
`ZeRO-2 <https://www.microsoft.com/en-us/research/blog/zero-2-deepspeed-shattering-barriers-of-deep-learning-speed-scale/>`_,
however the implementation is built from the ground up to be pytorch compatible and standalone.

Optimizer Sharded Training still utilizes Data Parallel Training under the hood, except optimizer state and gradients are sharded across GPUs.
This means the memory overhead per GPU is lower, as each GPU only has to maintain a section of your optimizer state and gradients.

The benefits are variable by model, but we've recorded up to a 25% memory reduction per GPU. Because of extremely efficient communication,

> these benefits in multi-GPU setups are almost free and throughput scales well with multi-node setups.

Reviewer: Are there any scenarios where this is not recommended?
